javalib/src/main/scala/java/util/Hashtable.scala

```diff
@@ -87,25 +87,87 @@ class Hashtable[K, V] private (inner: mutable.HashMap[Box[Any], V])
   override def toString(): String =
     inner.iterator.map(kv => "" + kv._1.inner + "=" + kv._2).mkString("{", ", ", "}")
 
-  def keySet(): ju.Set[K] =
-    inner.keySet.map(_.inner.asInstanceOf[K]).asJava
+  def keySet(): ju.Set[K] = {
+    new AbstractSet[K] {
+      def iterator(): Iterator[K] =
+        new EntrySetIterator().scalaOps.map(_.getKey())
+
+      def size(): Int = inner.size
+    }
+  }
 
   def entrySet(): ju.Set[ju.Map.Entry[K, V]] = {
-    class UnboxedEntry(
-        private[UnboxedEntry] val boxedEntry: ju.Map.Entry[Box[Any], V])
-        extends ju.Map.Entry[K, V] {
-      def getKey(): K = boxedEntry.getKey.inner.asInstanceOf[K]
-      def getValue(): V = boxedEntry.getValue
-      def setValue(value: V): V = boxedEntry.setValue(value)
-      override def equals(o: Any): Boolean = o match {
-        case o: UnboxedEntry => boxedEntry.equals(o.boxedEntry)
-        case _               => false
+    new AbstractSet[Map.Entry[K, V]] {
+      def iterator(): Iterator[Map.Entry[K, V]] =
+        new EntrySetIterator
+
+      def size(): Int = inner.size
+    }
+  }
+
+  private class EntrySetIterator extends Iterator[Map.Entry[K, V]] {
+    private var underlying: scala.collection.Iterator[(Box[Any], V)] =
+      Hashtable.this.inner.iterator
+
+    private var isSnapshot: Boolean = false
+
+    private var prev: Box[Any] = null
+
+    private def ensureSnapshot(): Unit = {
```

Comment:
I assume this logic is copied from Scala itself?

Reply:
More or less, but there are adaptations. I have added a comment.